hazelcast/src/main/java/com/hazelcast/concurrent/flakeidgen/FlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,200 @@
+     * <p>
+     * {@code 1483228800000} is the value {@code System.currentTimeMillis()} would return on
+     * 1.1.2017 0:00 GMT.
+     */
+    @SuppressWarnings("checkstyle:magicnumber")
+    static final long EPOCH_START = 1483228800000L + (1L << (BITS_TIMESTAMP - 1));
+    static final int BITS_NODE_ID = 16;
+
+    private static final int BITS_SEQUENCE = 6;
+
+    /** Assigned node ID or -1, if it is out of range */
+    private final int nodeId;
+    private final String name;
+    private final ILogger logger;
+
+    /**
+     * The highest timestamp|seq value returned so far. The value is not shifted to most significant bits.
+     */
+    private final AtomicLong generatedValue = new AtomicLong(Long.MIN_VALUE);
+
+    private volatile Member randomMember;
+    private AutoBatcher batcher;
+
+    /**
+     * Set of member UUIDs of which we know have node IDs out of range. These members are never again used
+     * to generate unique IDs, because this error is unrecoverable.
+     */
+    private final Set<String> outOfRangeMembers;
+
+    FlakeIdGeneratorProxy(String name, int nodeId, NodeEngine nodeEngine, FlakeIdGeneratorService service) {
+        super(nodeEngine, service);
+        this.name = name;
+        this.logger = nodeEngine.getLogger(getClass());
+
+        // If our nodeId is out of range, ID generation will be forwarded to another member.
+        // This is designed to provide functionality when the node ID overflows for new members but there are
+        // still some members with good node ID.
+        if ((nodeId & -1 << BITS_NODE_ID) != 0) {
+            this.nodeId = -1;
+            outOfRangeMembers = newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+            outOfRangeMembers.add(nodeEngine.getClusterService().getLocalMember().getUuid());
+            logger.severe("Node ID is out of range (" + nodeId + "), this member won't be able to generate IDs. "
+                    + "Cluster restart is recommended.");
+        } else {
+            this.nodeId = nodeId;
+            outOfRangeMembers = null;
+        }
+
+        FlakeIdGeneratorConfig config = nodeEngine.getConfig().findFlakeIdGeneratorConfig(getName());
+        batcher = new AutoBatcher(config.getPrefetchCount(), config.getPrefetchValidityMillis(),
+                new IFunction<Integer, IdBatch>() {
+                    @Override
+                    public IdBatch apply(Integer batchSize) {
+                        return newIdBatch(batchSize);
+                    }
+                });
+    }
+
+    @Override
+    public long newId() {
+        return batcher.newId();
+    }
+
+    @Override
+    public IdBatch newIdBatch(int batchSize) {
+        // local operation if we have valid node ID
+        if (nodeId >= 0) {
+            long base = newIdBaseLocal(batchSize);
+            return new IdBatch(base, 1 << BITS_NODE_ID, batchSize);
+        }
+
+        // remote call otherwise
+        while (true) {
```

Comment:
worst case scenario: if there is no node with a legitimate node id, we fall into an infinite loop here. We should stop looking for a legitimate member at some point, like, when all cluster members are put into the `outOfRangeMembers` set.

Reply:
sorry, I just noticed that this condition is checked in `getRandomMember()` method.